rdfproxy/mapper.py

```diff
@@ -0,0 +1,87 @@
+"""ModelBindingsMapper: Functionality for mapping binding maps to a Pydantic model."""
+
+from collections.abc import Iterator
+from typing import Any, Generic, get_args
+
+from pydantic import BaseModel
+from rdfproxy.utils._types import _TModelInstance
+from rdfproxy.utils.utils import (
+    _collect_values_from_bindings,
+    _get_group_by,
+    _get_key_from_metadata,
+    _list_basemodel_p,
+    _list_p,
+)
+
+
+class ModelBindingsMapper(Generic[_TModelInstance]):
+    """Utility class for mapping flat bindings to a (potentially nested) Pydantic model."""
+
+    def __init__(self, model: type[_TModelInstance], *bindings: dict):
+        self.model = model
+        self.bindings = bindings
+        self._contexts = []
+
+    def get_models(self) -> list[_TModelInstance]:
+        """Generate a list of (potentially nested) Pydantic models based on (flat) bindings."""
+        return self._get_unique_models(self.model, self.bindings)
+
+    def _get_unique_models(self, model, bindings):
+        """Call the mapping logic and collect unique and non-empty models."""
+        models = []
+        for _bindings in bindings:
+            _model = model(**dict(self._generate_binding_pairs(model, **_bindings)))
+
+            if any(_model.model_dump().values()) and (_model not in models):
+                models.append(_model)
+
+        return models
+
+    def _get_group_by(self, model, kwargs) -> str:
+        """Get the group_by value from a model and register it in self._contexts."""
+        group_by: str = _get_group_by(model, kwargs)
+
+        if group_by not in self._contexts:
+            self._contexts.append(group_by)
```

Comment:
Why not use a set?

Reply:
Because sets are unordered and an index 0 lookup on `self._context` is required.